### ISSUE TITLE
Segmentation fault when building ISO image with nixFlakes

### ISSUE THREAD

**Issue by rbvermaa** [NixOS/nix#4246]
**Describe the bug**

Segmentation fault when building ISO image with nixFlakes from release-20.09

**Steps To Reproduce**

To reproduce (on nixpkgs release-20.09):
```
nix-instantiate -A iso_minimal_new_kernel.x86_64-linux nixos/release.nix
```

**Expected behavior**

No segmentation fault

**nix (Nix) 2.4pre20201102_550e11f**

**Comment by roberth** [comment]
Probably #4178 which has been fixed in master. That time it wasn't reproduced by Eelco, so it's probably best if you try for yourself. You could use https://github.com/NixOS/nixpkgs/pull/102932 if you think that's more convenient.

**Comment by jonringer** [comment]
For others who land on this page.

Until there's a proper fix, you may try adding `GC_DONT_GC=1` to the command.

```
GC_DONT_GC=1 nix build ...
```

this seems to allow IFD to complete successfully for me

**Comment by pinpox** [comment]
> For others who land on this page.
> 
> Until there's a proper fix, you may try adding `GC_DONT_GC=1` to the command.
> 
> ```
> GC_DONT_GC=1 nix build ...
> ```
> 
> this seems to allow IFD to complete successfully for me

This "fixed" the problem for me aswell, but it seems more like a workaround. Any plans to get this fixed properly?

**Comment by jaen** [comment]
It has just now happened to me while building a non-ISO configuration, so it seems it's not limited to something that the ISO derivation is doing.

**Comment by tomberek** [comment]
This occurred with building an amazonImage.
- tried with flakes
- tried with legacy
GC_DONT_GC fixed it.
nix (Nix) 2.4pre20201205_a5d85d0

**Comment by ggreif** [comment]
Sorry for the linkspam above... I'll leave [this comment](https://github.com/dfinity/motoko/issues/2912#issuecomment-974556060) here, as it looks like there is still some lurking bug in 2.4.

**Comment by snajpa** [comment]
yes it is, even in 2.6... I'm being bit by this or something related, flakes aren't involved in my setup, it's a NIX_PATH override with heavy overlays... prefixing `nix-build` with `GC_DONT_GC=1` seems to avoid the segfault

**Comment by nrdxp** [comment]
hit a segfault in 2.11.1 that was resolved with the `GC_DONT_GC` var when building a devshell with some heavy IFD, so this bug is likely still around

**Comment by omnibs** [comment]
Same on 2.12.0, also resolved with `GC_DONT_GC`

**Comment by n8henrie** [comment]
Same, `Segmentation fault (core dumped)` building an aarch64 iso image. Works with `GC_DONT_GC`. Would love to learn more about how to debug these.

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.1-arch1-1, Arch Linux, noversion, rolling`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.12.0`
 - channels(root): `"nixpkgs-22.11"`
 - channels(n8henrie): `"nixpkgs-22.11"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/n8henrie/channels/nixpkgs`
```

**Comment by terlar** [comment]
I just hit this as well on a PR that updates my flake inputs. I could try to find out which updated input that caused the issue. This also only happens in GH Actions, running the same command locally had no issues.

**Comment by roberth** [comment]
Probably needs #7679. Is it a linux action or macos?

**Comment by terlar** [comment]
It was a Linux action

**Comment by terlar** [comment]
I have isolated the issue being triggered by nixpkgs upgrade:
https://github.com/terlar/nix-config/pull/125/commits/83ef162a8b14cc6170aa00f0ea1b78e9a2d0895f

I have not bisected anything to figure out exact at what point it broke. This breaks evaluation consistently for my GH action. Good to see a fix, will have to try it out when a new version is released.

Not sure if there are any hints in this information in order to make something small and reproducible.

**Comment by yannham** [comment]
I don't know if I'm hitting the same issue, but as @terlar, I get a segmentation fault on a `nix flake check` run by the CI of a PR updating flake inputs (generated by the `update_flake_lock` action) on a MacOS runner. The Linux runner got cancelled, so hard to say if it would be affected or not): https://github.com/tweag/nickel/actions/runs/4172595547/jobs/7223897501#step:5:23. Mentioning as a potential data point.
